File: app/src/main/cpp/board/board_test.cpp
/*
 * Copyright 2023 Michele Borassi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <gtest/gtest.h>
#include <gmock/gmock-matchers.h>
#include <gmock/gmock.h>
#include <regex>
#include "board.h"
#include "get_moves.h"

TEST(Board, Basic) {
  std::string board_string = "--------"
                             "--------"
                             "--------"
                             "--------"
                             "--------"
                             "--------"
                             "--------"
                             "------OX";
  Board b(board_string.c_str(), true);
  EXPECT_EQ(b, Board(1, 2));
  EXPECT_EQ(b.Player(), 1);
  EXPECT_EQ(b.Opponent(), 2);

  EXPECT_FALSE(b.IsEmpty(0));
  EXPECT_FALSE(b.IsEmpty(1));
  EXPECT_TRUE(b.IsEmpty(2));

  EXPECT_TRUE(b.IsPlayer(0));
  EXPECT_FALSE(b.IsPlayer(1));
  EXPECT_FALSE(b.IsPlayer(2));

  EXPECT_FALSE(b.IsOpponent(0));
  EXPECT_TRUE(b.IsOpponent(1));
  EXPECT_FALSE(b.IsOpponent(2));

  std::ostringstream ss;
  ss << b;
  EXPECT_EQ(std::regex_replace(ss.str(), std::regex("\\n"), ""), board_string);
}

TEST(Board, Sequence) {
  Board b("e6f4c3c4d3");
  Board b_copy("--------"
               "--------"
               "--XX----"
               "--OXOO--"
               "---XX---"
               "----X---"
               "--------"
               "--------", false);
  EXPECT_EQ(b, b_copy);
}

TEST(Board, BoardsFromSequenceEmpty) {
  std::vector<Board> previous;
  Board b("", &previous);
  EXPECT_THAT(previous, ::testing::ElementsAre());
}

TEST(Board, BoardsFromSequenceBase) {
  std::vector<Board> previous;
  Board b("e6f4c3c4d3", &previous);
  EXPECT_THAT(
      previous,
      ::testing::ElementsAre(
          Board(), Board("e6"), Board("e6f4"), Board("e6f4c3"),
          Board("e6f4c3c4"))
  );
}

TEST(Board, BoardsFromSequencePass) {
  std::vector<Board> previous;
  Board b("e6f6g6f4f5g7f7h7h8d6h6g8f8", &previous);
  EXPECT_THAT(
      previous,
      ::testing::ElementsAre(
          Board(""),
          Board("e6"),
          Board("e6f6"),
          Board("e6f6g6"),
          Board("e6f6g6f4"),
          Board("e6f6g6f4f5"),
          Board("e6f6g6f4f5g7"),
          Board("e6f6g6f4f5g7f7"),
          Board("e6f6g6f4f5g7f7h7"),
          Board("e6f6g6f4f5g7f7h7h8"),
          Board("e6f6g6f4f5g7f7h7h8d6"),
          Board("e6f6g6f4f5g7f7h7h8d6h6"),
          Board("e6f6g6f4f5g7f7h7h8d6h6g8"),
          Board(b.Opponent(), b.Player())
      )
  );
}

TEST(Board, BoardsFromSequencePass1) {
  std::vector<Board> previous;
  Board b("e6f6g6f4f5g7f7h7h8d6h6g8f8g5", &previous);
  Board result_after_pass = Board("e6f6g6f4f5g7f7h7h8d6h6g8f8");
  EXPECT_THAT(
      previous,
      ::testing::ElementsAre(
          Board(""),
          Board("e6"),
          Board("e6f6"),
          Board("e6f6g6"),
          Board("e6f6g6f4"),
          Board("e6f6g6f4f5"),
          Board("e6f6g6f4f5g7"),
          Board("e6f6g6f4f5g7f7"),
          Board("e6f6g6f4f5g7f7h7"),
          Board("e6f6g6f4f5g7f7h7h8"),
          Board("e6f6g6f4f5g7f7h7h8d6"),
          Board("e6f6g6f4f5g7f7h7h8d6h6"),
          Board("e6f6g6f4f5g7f7h7h8d6h6g8"),
          Board(result_after_pass.Opponent(), result_after_pass.Player()),
          Board("e6f6g6f4f5g7f7h7h8d6h6g8f8")
      )
  );
}

TEST(Board, Unique) {
  for (int i = 0; i < 1000; ++i) {
    Board b = RandomBoard();
    for (Board t : b.AllTranspositions()) {
      EXPECT_EQ(b.Unique(), t.Unique());
    }
  }
}

TEST(Board, UndoMove) {
  for (int i = 0; i < 1000; ++i) {
    const Board b = RandomBoard();
    const auto flips = GetAllMoves(b.Player(), b.Opponent());
    if (flips.size() == 0) {
      continue;
    }
    BitPattern flip = flips[rand() % flips.size()];
    Board b_copy = b.Next(flip);
    b_copy.UndoMove(flip, b.MoveFromFlip(flip));
    ASSERT_EQ(b, b_copy);
  }
}

TEST(SerializedBoard, Serialize) {
  for (int i = 0; i < 1000; ++i) {
    Board b = RandomBoard();
    const auto serialized = b.Serialize();
    EXPECT_EQ(b, Board::Deserialize(serialized.begin()));
  }
}

File: app/src/test/java/board/BoardTest.java
// Copyright 2019 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package board;

import java.util.ArrayList;
import static org.junit.Assert.*;

import java.util.Arrays;
import java.util.HashSet;
import java.util.List;

import org.junit.Test;

public class BoardTest {
  
  public static <T> void assertEqualsIgnoreOrder(T[] list1, T[] list2) {
    HashSet<T> tmp1 = new HashSet<>(Arrays.asList(list1));
    HashSet<T> tmp2 = new HashSet<>(Arrays.asList(list2));
    assertEquals(tmp1, tmp2);
  }

  public static <T> void assertListEqualsIgnoreOrder(List<T> list1, List<T> list2) {
    assertEquals(list1.size(), list2.size());
    HashSet<T> tmp1 = new HashSet<>(list1);
    HashSet<T> tmp2 = new HashSet<>(list2);
    assertEquals(tmp1, tmp2);
  }
  
  public static String randomBoardString() {
    String boardString = "";
    for (int j = 0; j < 64; j++) {
      int rand = (int) (Math.random() * 3);
      if (rand == 0) {
        boardString += "-";
      } else if (rand == 1) {
        boardString += "O";
      } else {
        boardString += "X";
      }
      if (j % 8 == 7) {
        boardString += "\n";
      }
    }
    return boardString;
  }

  public static Board randomBoard() {
    return new Board(randomBoardString(), Math.random() > 0.5);
  }

  @Test
  public void boardFromString() {
    assertEquals(new Board("e6"), new Board("e6"));
    assertEquals(new Board("e6f4"), new Board("e6f4"));
  }

  @Test
  public void testBoardFromString() {
    String boardString = "--------\n"
                       + "--------\n" 
                       + "--------\n"
                       + "---OX---\n"
                       + "---XO---\n"
                       + "--------\n"
                       + "--------\n"
                       + "--------\n";
    Board board = new Board(boardString, true);
    assertEquals(new Board(
            0b0000000000000000000000000000100000010000000000000000000000000000L,
            0b0000000000000000000000000001000000001000000000000000000000000000L), board);
  }

  int countOccurrencesInString(String s, char c) {
    int result = 0;
    for (char i : s.toCharArray()) {
      if (i == c) {
        result++;
      }
    }
    return result;
  }
  
  @Test
  public void testCountDisks() {
    for (int i = 0; i < 10000; i++) {
      Board b = randomBoard();

      int playerDisks = countOccurrencesInString(b.toString(), 'X');
      assertEquals(b.getPlayerDisks(), playerDisks);
      int opponentDisks = countOccurrencesInString(b.toString(), 'O');
      assertEquals(b.getOpponentDisks(), opponentDisks);
      int emptySquares = countOccurrencesInString(b.toString(), '-');
      assertEquals(b.getEmptySquares(), emptySquares);
      int fullSquares = playerDisks + opponentDisks;
      assertEquals(b.getFullSquares(), fullSquares);
    }
  }
}
